Automattic/jetmon
# Enhancement: Run the main loop every minute to allow variable intervals

Currently, the parent's main loop ran every five minutes. This PR adds the ability to run the loop each minute while checking sites on an interval set by the `check_interval` value in the database. To enable this new functionality, the `USE_VARIABLE_CHECK_INTERVALS` value must be set to `true` in the `config/config.json` file (see `config/config-sample.json` for an example of how to add this to an existing config file).

This functionality is implemented by a new query with a calculation that uses a site row's ID to spread out requests across the span of their check interval. When the query is run each minute, a site with a one minute check interval will return on every query for its bucket whereas a site with a five minute check interval will only return once every five queries.

This PR has the following goals:

1. Ensure that sites with a 1-minute check interval can be scanned each minute.
2. Ensure that sites that have a check interval greater than 1-minute spread out across requests each minute rather than clumping into requests on specific minutes. For example, we don't want all sites with the default 5-minute check interval to only scan at 5, 10, 15, 20, etc minutes.
3. Allow this behavior to be enabled and disabled via a config change to allow for easily reverting the production systems in case a problem is discovered after deployment.

### Testing Instructions

1. Ensure that the test Jetmon database includes sites with check intervals other than 5 minutes. Ensure that at least one site is set to a check interval of 1 minute. Note: Only check intervals that are divisors of 60 are valid (1, 2, 3, 4, 5, 6, 10, 12, 15, 20, 30 and 60).
2. Ensure that the test Jetmon database includes multiple sites that have a check interval greater than 1 minute and share the same check interval. For best results, add at least two test sites with the same check interval greater than 1 using back-to-back inserts so that their ID's are sequential.
4. Set `USE_VARIABLE_CHECK_INTERVALS` to `false` in `config/config.json` (check `config/config-sample.json` for an example).
5. Set `MIN_TIME_BETWEEN_ROUNDS_SEC` to `300` to ensure that sites are scanned every five minutes.
6. Apply this PR.
7. Run Jetmon.
8. Verify that all the sites are scanned every five minutes. This includes sites with different check intervals, such as sites with 1-minute check intervals.
9. Stop Jetmon.
10. Set `USE_VARIABLE_CHECK_INTERVALS` to `true` in `config/config.json`.
11. Run Jetmon.
12. Verify that the sites in the test Jetmon database are scanned at the check intervals corresponding with the `check_interval` value for the site in the database.
13. Verify that sites with a 5-minute check interval have their requests spread out across multiple rounds during the five minute check interval rather than all scanning during the same round. For example, if there are 100 sites with a 5-minute check interval, approximately 20 of the sites should be scanned each minute with all 100 being scanned across a five minute span. (Note: the site's site ID determines how each site is divided into these one minute groupings, so some groups may be larger or smaller than 20).
14. Verify that sites with check intervals other than 1 minute and 5 minutes scan at the expected intervals.
